tls/s2n_resume.c

```diff
@@ -769,187 +769,147 @@ struct s2n_ticket_key *s2n_find_ticket_key(struct s2n_config *config, const uint
-
+    /* Initialize blob to be decrypted */
     struct s2n_blob en_blob = { 0 };
     uint32_t en_blob_size = s2n_stuffer_data_available(from);
     uint8_t *en_blob_data = s2n_stuffer_raw_read(from, en_blob_size);
-    POSIX_ENSURE_REF(en_blob_data);
-    POSIX_GUARD(s2n_blob_init(&en_blob, en_blob_data, en_blob_size));
-    POSIX_GUARD(s2n_aes256_gcm.io.aead.decrypt(&aes_ticket_key, &iv, &aad_blob, &en_blob, &en_blob));
+    RESULT_ENSURE_REF(en_blob_data);
+    RESULT_GUARD_POSIX(s2n_blob_init(&en_blob, en_blob_data, en_blob_size));
 
+    RESULT_GUARD_POSIX(s2n_aes256_gcm.io.aead.decrypt(&aes_ticket_key, &iv, &aad_blob, &en_blob, &en_blob));
+    RESULT_GUARD(s2n_aes256_gcm.destroy_key(&aes_ticket_key));
+
+    /* Parse decrypted state */
     struct s2n_blob state_blob = { 0 };
     uint32_t state_blob_size = en_blob_size - S2N_TLS_GCM_TAG_LEN;
-    POSIX_GUARD(s2n_blob_init(&state_blob, en_blob.data, state_blob_size));
+    RESULT_GUARD_POSIX(s2n_blob_init(&state_blob, en_blob.data, state_blob_size));
     struct s2n_stuffer state_stuffer = { 0 };
-    POSIX_GUARD(s2n_stuffer_init(&state_stuffer, &state_blob));
-    POSIX_GUARD(s2n_stuffer_skip_write(&state_stuffer, state_blob_size));
-    POSIX_GUARD_RESULT(s2n_deserialize_resumption_state(conn, &from->blob, &state_stuffer));
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&state_stuffer, &state_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(&state_stuffer, state_blob_size));
+    RESULT_GUARD(s2n_deserialize_resumption_state(conn, &from->blob, &state_stuffer));
 
-    if (s2n_connection_get_protocol_version(conn) >= S2N_TLS13) {
-        return S2N_SUCCESS;
+    if (s2n_connection_get_protocol_version(conn) >= S2N_TLS13 || conn->config->use_session_cache) {
+        return S2N_RESULT_OK;
     }
 
     /* A new key is assigned for the ticket if the key used to encrypt current ticket is expired */
     uint64_t now = 0;
-    POSIX_GUARD_RESULT(s2n_config_wall_clock(conn->config, &now));
+    RESULT_GUARD(s2n_config_wall_clock(conn->config, &now));
     if (now >= key->intro_timestamp + conn->config->encrypt_decrypt_key_lifetime_in_nanos) {
         if (s2n_result_is_ok(s2n_config_is_encrypt_key_available(conn->config))) {
             conn->session_ticket_status = S2N_NEW_TICKET;
-            POSIX_GUARD_RESULT(s2n_handshake_type_set_tls12_flag(conn, WITH_SESSION_TICKET));
+            RESULT_GUARD(s2n_handshake_type_set_tls12_flag(conn, WITH_SESSION_TICKET));
         }
     }
-    return S2N_SUCCESS;
+    return S2N_RESULT_OK;
 }
 
 int s2n_encrypt_session_cache(struct s2n_connection *conn, struct s2n_stuffer *to)
 {
-    return s2n_encrypt_session_ticket(conn, to);
+    POSIX_GUARD_RESULT(s2n_resume_encrypt_session_ticket(conn, to));
+    return S2N_SUCCESS;
 }
 
-int s2n_decrypt_session_cache(struct s2n_connection *conn, struct s2n_stuffer *from)
+int s2n_resume_decrypt_session_cache(struct s2n_connection *conn, struct s2n_stuffer *from)
 {
```

Comment:
To make sure we're not losing anything, could you do a diff of s2n_resume_decrypt_session_cache and (the old implementation of) s2n_resume_decrypt_session_ticket? Probably just stick it in a gist

Reply:
https://gist.github.com/maddeleine/284a6f2d7769cc05b74e509e61fa18a6
So there's only one functional thing that concerns me when comparing the two older versions. If you look at the diff, the session cache function calls the deserialization function like so: 
`s2n_deserialize_resumption_state(conn, NULL, &state_stuffer)`
versus the session ticket function calls it like this: 
`s2n_deserialize_resumption_state(conn, &from->blob, &state_stuffer)`

The middle parameter is used corresponds to (in tls13) the psk identity field and the session ticket that gets sent in the client_hello (in tls12). 

So, I don't actually think this change would "functionally" affect the session cache feature. But it is a change nonetheless. I think I'm going to just keep the decrypt_session_cache() and decrypt_session_ticket() functions separate for now.

Reply:
You might want to add that gist to the description for your second reviewer to review.